pkg/shardddl/optimism/lock.go

```diff
@@ -149,13 +157,19 @@ func (l *Lock) TrySync(info Info, tts []TargetTable) (newDDLs []string, err erro
 		return ddls, terror.ErrMasterInconsistentOptimisticDDLsAndInfo.Generate(len(ddls), len(newTIs))
 	}
 
+	// should not happen
+	if info.TableInfoBefore == nil {
+		return ddls, terror.ErrMasterOptimisticTableInfoBeforeNotExist.Generate(ddls)
+	}
+	oldTable := schemacmp.Encode(info.TableInfoBefore)
 	// handle the case where <callerSource, callerSchema, callerTable>
 	// is not in old source tables and current new source tables.
 	// duplicate append is not a problem.
 	tts = append(tts, newTargetTable(l.Task, callerSource, l.DownSchema, l.DownTable,
 		map[string]map[string]struct{}{callerSchema: {callerTable: struct{}{}}}))
 	// add any new source tables.
 	l.addTables(tts)
+	l.receiveTable(callerSource, callerSchema, callerTable, oldTable)
 	if val, ok := l.versions[callerSource][callerSchema][callerTable]; !ok || val < infoVersion {
 		l.versions[callerSource][callerSchema][callerTable] = infoVersion
 	}
```

Comment:
now we changed the `l.tables` by `l.receiveTable`, so in line 179 `oldJoined` is not a result of "join every l.tables" 🤔

Reply:
Did you mean join all the tables for old joined?

Reply:
I'm just not sure if we depend on "old joined should be joined result of all l.tables", or "old joined of previous TrySync and new joined of next TrySync should keep consistent"

Reply:
If we assume user start-task with same table info in all upstream, tableInfoBefore should be same as oldJoined. Otherwise, the oldjoined is not collect and we may keep it and hope later join may return an error? cc @lichunzhu

Reply:
If dm-master/dm-worker meets a restart the table info in all upstream is more likely to be different. What's the problem here? I don't understand this clearly.

Reply:
I think the tableInfoBefore is same as l.tables[source][schema][table], otherwise it's a bug or idempotent retry.

Reply:
But if the tableInfoBefore not equal l.tables[source][schema][table], that means the l.tables[source][schema][table] and oldjoined may wrong, do we need join all the tables with tableInfoBefore as oldjoined?

Reply:
> But if the tableInfoBefore not equal l.tables[source][schema][table], that means the l.tables[source][schema][table] and oldjoined may wrong, do we need join all the tables with tableInfoBefore as oldjoined?

I think carefully about this situation. Could you help me check if I'm wrong?

When this happens, it means we add a table whose table schema is not equal to joined table info now. When we init all the tables, we have three situations:
1. This task is new. The pre-check will assure the correctness of the tables' schemas.
2. This dm cluster recovers from a failure. The lock is synced now and all the tables have the same schema. **If we add a table with a different schema, we should report an error.**
3. This dm cluster recovers from a failure. The lock is unsynced now. Some workers didn't put their lock Info into etcd before the restart. These tables' `l.tables` will be set to `l.joined` at first, and if it's different from the `tableInfoBefore` received later I think it's acceptable. Am I right?

Reply:
So you mean if we get a info which we have already receive the table info before, if their schema is different, we should report an error?

Reply:
Yes. But I think we shouldn't report an error in situation 3.

Reply:
> Some workers didn't put their lock Info into etcd before the restart. These tables' l.tables will be set to l.joined at first, and if it's different from the tableInfoBefore received later

That means the table haven't received. `l.received[x][x][x]=false`

In unit test, we have some idempotent TrySync, how to deal with them?

Reply:
> Some workers didn't put their lock Info into etcd before the restart. These tables' l.tables will be set to l.joined at first

That means the table haven't been received. `l.received[x][x][x]==false`

> In unit test, we have some idempotent TrySync, how do we deal with the case? 🤔